[MRG+1] Dropping python 2.6 support

lesteve opened the pull request:
0.19 is the first release without Python 2.6 support. Related: #7522.

The `git grep` I used to try to remove Python 2.6 code:
```
git grep -iP 'py(thon).*2\.6' | grep -v externals
```

lesteve committed:
Remove Python 2.6 support

Some details about some slightly orthogonal changes:
* Note about cheking safely for nan is likely not valid any more (commit
  introducing it is c80ca91b)
* scipy.linalg.qr econ parameter removed since scipy 0.9 in favour of
  mode='economic'
* Remove unnecessary libgfortran in conda create command

lesteve committed:
Putative fix by setting the random seed

lesteve commented on sklearn/gaussian_process/gaussian_process.py:
```diff
@@ -618,14 +618,7 @@ def reduced_likelihood_function(self, theta=None):
         # Get generalized least squares solution
         Ft = linalg.solve_triangular(C, F, lower=True)
-        try:
-            Q, G = linalg.qr(Ft, econ=True)
```
econ=True removed in scipy=0.9 which matches our minimum scipy version. From 
https://github.com/scipy/scipy/blob/master/doc/release/0.9.0-notes.rst#removed-features.

lesteve commented on sklearn/metrics/classification.py:
```diff
@@ -431,10 +431,7 @@ def jaccard_similarity_score(y_true, y_pred, normalize=True,
-            # the jaccard to 1: lim_{x->0} x/x = 1
-            # Note with py2.6 and np 1.3: we can't check safely for nan.
-            score[pred_or_true == 0.0] = 1.0
+            score[np.isnan(score)] = 1.0
```
Bold move I know. The commit introducing this change is c80ca91b and no tests were added for this edge case along with the change. Does anyone remember a problem like this?

amueller replied:
can't we do the same thing here we do in places like ``StandardScaler``?

amueller replied:
Modifying the denominator, so setting ``pred_or_true[pred_or_true == 0] = 1``. Then we don't need to catch the error, either.

amueller replied:
Wait, sorry, we want score to be one, not zero. We should really have a standard way of dividing by zero, it comes up soo often. what's wrong with the current solution though? I find it cleaner than the change.

lesteve replied:
Fine then I'll leave the previous solution and just remove the comment. Given the comment I thought something in an old python + numpy was preventing to just check for NaNs.

lesteve commented on sklearn/neighbors/tests/test_approximate.py:
```diff
@@ -229,7 +229,8 @@ def test_radius_neighbors_boundary_handling():
     # exact results on this toy dataset.
-    lsfh = LSHForest(min_hash_match=0, n_candidates=n_points).fit(X)
+    lsfh = LSHForest(min_hash_match=0, n_candidates=n_points,
+                     random_state=42).fit(X)
```
I don't know why this is needed ... but it is needed without it you get an error consistently like [this](https://travis-ci.org/lesteve/scikit-learn/jobs/176360295#L1989) if you run the tests with `nosetests sklearn`. The error can not be reproduced only with running a single test function or the single file with `nosetests sklearn/neighbors/tests/test_approximate.py` and I tried in a loop 100 times. Still trying to get to the bottom of this ...

jnothman replied:
The implication being that in removing one of the fixes you changed behaviour?

amueller replied:
well this should be here anyhow...

lesteve commented on .travis.yml:
```diff
@@ -29,7 +29,7 @@ env:
       COVERAGE=true
     # This environment tests the oldest supported anaconda env
-    - DISTRIB="conda" PYTHON_VERSION="2.6" INSTALL_MKL="false"
+    - DISTRIB="conda" PYTHON_VERSION="2.7" INSTALL_MKL="false"
```
Still keeping this build to test oldest versions available through conda.

lesteve reviewed (commented):
A few comments to help the review

jnothman commented on sklearn/utils/tests/test_estimator_checks.py:
```diff
@@ -84,7 +84,7 @@ def test_check_estimator():
     assert_raises_regex(AttributeError, msg, check_estimator, BaseEstimator)
     # check that fit does input validation
-    msg = "TypeError not raised by fit"
+    msg = "TypeError not raised"
```
huh?

amueller replied:
I second @jnothman's concern ;)

jnothman commented:
Thanks for this, @lesteve!

amueller commented on build_tools/travis/install.sh:
```diff
@@ -53,23 +53,17 @@ if [[ "$DISTRIB" == "conda" ]]; then
     if [[ "$INSTALL_MKL" == "true" ]]; then
         conda create -n testenv --yes python=$PYTHON_VERSION pip nose \
             numpy=$NUMPY_VERSION scipy=$SCIPY_VERSION numpy scipy \
-            libgfortran mkl flake8 \
```
what's the reasoning with libgfortran?

lesteve replied:
I think at one point it was needed explicitly because dependencies were not correctly defined in the conda metadata. This has long been fixed.

amueller reviewed (approved):
LGTM apart from minor comments.

lesteve committed:
Revert unintended change

lesteve committed:
Reinstate previous logic for checking for NaNs

lesteve committed:
Reinstate change in error message

Error messages from Python 2.7 assertRegexp does not contain the
function name, in contrast with Python 3 assertRegex

amueller commented:
Hm do you keep force-pushing? I can't find your comments :-/

amueller commented on sklearn/utils/testing.py:
```diff
@@ -77,85 +77,27 @@
-                                  callable_obj.__name__))
-
+    # Python 2.7
+    assert_raises_regex = _dummy.assertRaisesRegexp
```
Hm or leave the old one in and not change behavior?

lesteve commented:
> Hm do you keep force-pushing? I can't find your comments :-/

I don't think force pushing will fix it. In case you are using firefox I have a GreaseMonkey script to open all outdated diff comments (updated it recently): https://gist.github.com/lesteve/b4ef29bccd42b354a834#file-github_outdated_diff-user-js. I have set the shortcut to Alt-P because that happen to be one of the shortcut not taken by the browser.

You can probably change it just a tiny bit and copy and paste into the console or turn it into a bookmarklet if you wish. For some reason I can't remember I never managed to get the latter to work.

jnothman commented:
LGTM

amueller commented:
LGTM, thanks :)

jnothman commented:
Yay!

On 24 November 2016 at 09:11, Andreas Mueller <notifications@github.com>
wrote:

> LGTM, thanks :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7890#issuecomment-262642439>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz66_G0bREc7kMvz29iTCrampnps26ks5rBLn-gaJpZM4Kz2fT>
> .
>
